lib/src/null_safety_analysis.dart

```diff
@@ -0,0 +1,165 @@
+  /// doesn't parse.
+  analysisFailed,
+}
+
+class NullSafetyAnalysis {
+  final SystemCache _systemCache;
+  final Map<PackageId, Future<NullSafetyCompliance>>
+      _packageInternallyGoodCache = {};
+  static final _firstVersionWithNullSafety = Version.parse('2.10.0');
+
+  NullSafetyAnalysis(SystemCache systemCache) : _systemCache = systemCache;
+
+  Future<NullSafetyCompliance> nullSafetyCompliance(PackageId packageId) async {
+    final result = await _nullSafetyCompliance(packageId);
+    return result;
+  }
+
+  /// Returns true if package version [packageId] and all its non-dev
+  /// dependencies (transitively) have a language version >= 2.10, and no files
+  /// in lib/ of  these packages opt out to a pre-2.10 language version.
+  ///
+  /// This will do a full resolution of that package's import graph, and also
+  /// download the package and all dependencies into [cache].
+  Future<NullSafetyCompliance> _nullSafetyCompliance(
+      PackageId packageId) async {
+    return await withTempDir((tempPath) async {
+      final importingPubspec = {
+        'name': '${packageId.name}_importer',
+        'dependencies': {
+          packageId.name: {
+            packageId.source.name: packageId.source is PathSource
+                ? packageId.description['path']
+                : packageId.description,
+            'version': packageId.version.toString(),
+          }
+        }
+      };
+
+      writeTextFile(
+          path.join(tempPath, 'pubspec.yaml'), json.encode(importingPubspec));
+      final importingEntrypoint = Entrypoint(tempPath, _systemCache);
+      SolveResult result;
+      try {
+        result = await resolveVersions(
+          SolveType.GET,
+          _systemCache,
+          importingEntrypoint.root,
+        );
+      } on SolveFailure {
+        return NullSafetyCompliance.analysisFailed;
+      }
+
+      final analysisSession = ContextBuilder()
+          .createContext(
+            contextRoot: ContextLocator().locateRoots(
+              includedPaths: [tempPath],
+            ).first,
+          )
+          .currentSession;
+
+      var allPackagesGood = true;
+      for (final dependencyId in result.packages) {
+        if (dependencyId.name == importingEntrypoint.root.name) continue;
+
+        final packageInternallyGood = await _packageInternallyGoodCache
+            .putIfAbsent(dependencyId, () async {
+          final boundSource = dependencyId.source.bind(_systemCache);
+          final pubspec = await boundSource.describe(dependencyId);
+          final languageVersion = _languageVersion(pubspec);
+          if (languageVersion == null ||
+              languageVersion < _firstVersionWithNullSafety) {
+            return NullSafetyCompliance.notCompliant;
+          }
+
+          if (boundSource is CachedSource) {
+            /// TODO(sigurdm): Should we set metadata here?
+            await boundSource.downloadToSystemCache(dependencyId);
```

Comment:
Since you put a future into `_packageInternallyGoodCache`, you can't call `downloadToSystemCache´ concurrently, so maybe all of my concerns are mute :D

Reply:
Yes!